File: agent/bot.py
"""Telegram bot handlers for resource curation."""
import re
import os
import httpx
from telegram import Update
from telegram.ext import (
    Application,
    CommandHandler,
    MessageHandler,
    ContextTypes,
    filters,
)
import api_client as api
from parser import scrape_og, pick_category

OWNER_ID = int(os.environ.get("TELEGRAM_OWNER_ID", "0"))

URL_RE = re.compile(r"https?://\S+")


def _is_owner(update: Update) -> bool:
    return OWNER_ID == 0 or update.effective_user.id == OWNER_ID


# ── URL handler ────────────────────────────────────────────────────────────────

async def handle_url(update: Update, ctx: ContextTypes.DEFAULT_TYPE) -> None:
    text = update.message.text or ""
    match = URL_RE.search(text)
    if not match:
        return
    url = match.group(0).rstrip(".,)")

    if not _is_owner(update):
        await update.message.reply_text("Not authorised.")
        return

    await update.message.reply_text("⏳ Fetching…")

    try:
        og = await scrape_og(url)
        title = og["title"]
        description = og["description"]

        cats = await api.get_categories()
        slug = await pick_category(url, title, description, cats)

        cat_name = next((c["name"] for c in cats if c["slug"] == slug), slug)

        result = await api.post_resource(url, title, description, slug)
        if result.get("success"):
            resource_id = result["data"]["id"]
            await update.message.reply_text(
                f'✅ Added to *{cat_name}* (#{resource_id})\n"{title}"',
                parse_mode="Markdown",
            )
        else:
            await update.message.reply_text(f"❌ Error: {result.get('error', 'unknown')}")
    except httpx.ConnectError:
        await update.message.reply_text("❌ Board unavailable — try again in a moment.")
    except Exception as e:
        await update.message.reply_text(f"❌ Unexpected error: {type(e).__name__}")


# ── Commands ───────────────────────────────────────────────────────────────────

async def cmd_list(update: Update, ctx: ContextTypes.DEFAULT_TYPE) -> None:
    category = ctx.args[0] if ctx.args else None
    items = await api.get_resources(category=category, limit=10)
    if not items:
        await update.message.reply_text("No saved links found.")
        return
    lines = [f"*Recent links{' in ' + category if category else ''}:*"]
    for r in items:
        lines.append(f"#{r['id']} [{r['title']}]({r['url']}) — _{r['category']['name']}_")
    await update.message.reply_text("\n".join(lines), parse_mode="Markdown", disable_web_page_preview=True)


async def cmd_delete(update: Update, ctx: ContextTypes.DEFAULT_TYPE) -> None:
    if not _is_owner(update):
        await update.message.reply_text("Not authorised.")
        return
    if not ctx.args or not ctx.args[0].isdigit():
        await update.message.reply_text("Usage: /delete <id>")
        return
    ok = await api.delete_resource(int(ctx.args[0]))
    await update.message.reply_text("✅ Deleted." if ok else "❌ Not found or error.")


async def cmd_search(update: Update, ctx: ContextTypes.DEFAULT_TYPE) -> None:
    query = " ".join(ctx.args)
    if not query:
        await update.message.reply_text("Usage: /search <query>")
        return
    items = await api.search_resources(query)
    if not items:
        await update.message.reply_text("No results.")
        return
    lines = [f"*Results for '{query}':*"]
    for r in items:
        lines.append(f"#{r['id']} [{r['title']}]({r['url']}) — _{r['category']['name']}_")
    await update.message.reply_text("\n".join(lines), parse_mode="Markdown", disable_web_page_preview=True)


async def cmd_categories(update: Update, ctx: ContextTypes.DEFAULT_TYPE) -> None:
    cats = await api.get_categories()
    lines = ["*Categories:*"] + [f"• `{c['slug']}` — {c['name']}" for c in cats]
    await update.message.reply_text("\n".join(lines), parse_mode="Markdown")


async def cmd_addcategory(update: Update, ctx: ContextTypes.DEFAULT_TYPE) -> None:
    if not _is_owner(update):
        await update.message.reply_text("Not authorised.")
        return
    if len(ctx.args) < 2:
        await update.message.reply_text("Usage: /addcategory <slug> <Name>")
        return
    slug = ctx.args[0].lower()
    name = " ".join(ctx.args[1:])
    result = await api.add_category(name, slug)
    if result.get("success"):
        await update.message.reply_text(f'✅ Added category *{name}* (`{slug}`)', parse_mode="Markdown")
    else:
        await update.message.reply_text(f"❌ {result.get('error', 'Error')}")


async def cmd_start(update: Update, _: ContextTypes.DEFAULT_TYPE) -> None:
    await update.message.reply_text(
        "📚 *Curator Board bot*\n\nSend any URL to save it.\n\n"
        "/list [category] — recent links\n"
        "/search <query> — search\n"
        "/delete <id> — remove\n"
        "/categories — list categories\n"
        "/addcategory <slug> <Name> — add category",
        parse_mode="Markdown",
    )


# ── App builder ────────────────────────────────────────────────────────────────

def build_app(token: str) -> Application:
    app = Application.builder().token(token).build()
    app.add_handler(CommandHandler("start", cmd_start))
    app.add_handler(CommandHandler("list", cmd_list))
    app.add_handler(CommandHandler("delete", cmd_delete))
    app.add_handler(CommandHandler("search", cmd_search))
    app.add_handler(CommandHandler("categories", cmd_categories))
    app.add_handler(CommandHandler("addcategory", cmd_addcategory))
    app.add_handler(MessageHandler(filters.TEXT & ~filters.COMMAND, handle_url))
    return app

File: agent/bot.ts
import { Bot, Context } from "grammy";
import {
  addCategory,
  deleteResource,
  getCategories,
  getResources,
  postResource,
  searchResources,
} from "./api-client";
import { pickCategory, scrapeOg } from "./parser";

const OWNER_ID = Number(process.env.TELEGRAM_OWNER_ID ?? "0");
const URL_RE = /(?:https?:\/\/\S+|(?:[a-zA-Z0-9-]+\.)+[a-zA-Z]{2,}\S*)/;

function isOwner(ctx: Context): boolean {
  return OWNER_ID === 0 || ctx.from?.id === OWNER_ID;
}

function parseArgs(text: string): string[] {
  return text.trim().split(/\s+/).slice(1);
}

function escapeHtml(value: string): string {
  return value
    .replaceAll("&", "&amp;")
    .replaceAll("<", "&lt;")
    .replaceAll(">", "&gt;")
    .replaceAll('"', "&quot;")
    .replaceAll("'", "&#39;");
}

async function handleUrl(ctx: Context): Promise<void> {
  const text = ctx.message?.text ?? "";
  const match = text.match(URL_RE);
  if (!match) {
    return;
  }

  const raw = match[0].replace(/[.,)]+$/, "");
  const url = /^https?:\/\//i.test(raw) ? raw : `https://${raw}`;
  if (!isOwner(ctx)) {
    await ctx.reply("Not authorised.");
    return;
  }

  await ctx.reply("⏳ Fetching…");

  try {
    const og = await scrapeOg(url);
    const categories = await getCategories();
    const slug = await pickCategory(url, og.title, og.description, categories);
    const categoryName = categories.find((category) => category.slug === slug)?.name ?? slug;
    const result = await postResource(url, og.title, og.description, slug);

    if (result.success && result.data) {
      await ctx.reply(
        `✅ Added to <b>${escapeHtml(categoryName)}</b> (#${result.data.id})\n"${escapeHtml(og.title)}"`,
        { parse_mode: "HTML" },
      );
      return;
    }

    await ctx.reply(`❌ Error: ${String(result.error ?? "unknown")}`);
  } catch (error) {
    if (error instanceof TypeError) {
      await ctx.reply("❌ Board unavailable — try again in a moment.");
      return;
    }

    await ctx.reply(`❌ Unexpected error: ${error instanceof Error ? error.name : "Error"}`);
  }
}

async function cmdList(ctx: Context): Promise<void> {
  const text = ctx.message?.text ?? "";
  const [category] = parseArgs(text);
  const items = await getResources(category, 10);

  if (items.length === 0) {
    await ctx.reply("No saved links found.");
    return;
  }

  const lines = [`<b>Recent links${category ? ` in ${escapeHtml(category)}` : ""}:</b>`];
  for (const item of items) {
    lines.push(
      `#${item.id} <a href="${item.url}">${escapeHtml(item.title)}</a> — <i>${escapeHtml(item.category.name)}</i>`,
    );
  }

  await ctx.reply(lines.join("\n"), {
    parse_mode: "HTML",
    link_preview_options: { is_disabled: true },
  });
}

async function cmdDelete(ctx: Context): Promise<void> {
  if (!isOwner(ctx)) {
    await ctx.reply("Not authorised.");
    return;
  }

  const text = ctx.message?.text ?? "";
  const [id] = parseArgs(text);
  if (!id || !/^\d+$/.test(id)) {
    await ctx.reply("Usage: /delete <id>");
    return;
  }

  const ok = await deleteResource(Number(id));
  await ctx.reply(ok ? "✅ Deleted." : "❌ Not found or error.");
}

async function cmdSearch(ctx: Context): Promise<void> {
  const text = ctx.message?.text ?? "";
  const query = parseArgs(text).join(" ");
  if (!query) {
    await ctx.reply("Usage: /search <query>");
    return;
  }

  const items = await searchResources(query);
  if (items.length === 0) {
    await ctx.reply("No results.");
    return;
  }

  const lines = [`<b>Results for '${escapeHtml(query)}':</b>`];
  for (const item of items) {
    lines.push(
      `#${item.id} <a href="${item.url}">${escapeHtml(item.title)}</a> — <i>${escapeHtml(item.category.name)}</i>`,
    );
  }

  await ctx.reply(lines.join("\n"), {
    parse_mode: "HTML",
    link_preview_options: { is_disabled: true },
  });
}

async function cmdCategories(ctx: Context): Promise<void> {
  const categories = await getCategories();
  const lines = ["<b>Categories:</b>"];
  for (const category of categories) {
    lines.push(`• <code>${escapeHtml(category.slug)}</code> — ${escapeHtml(category.name)}`);
  }
  await ctx.reply(lines.join("\n"), { parse_mode: "HTML" });
}

async function cmdAddCategory(ctx: Context): Promise<void> {
  if (!isOwner(ctx)) {
    await ctx.reply("Not authorised.");
    return;
  }

  const text = ctx.message?.text ?? "";
  const args = parseArgs(text);
  if (args.length < 2) {
    await ctx.reply("Usage: /addcategory <slug> <Name>");
    return;
  }

  const slug = args[0].toLowerCase();
  const name = args.slice(1).join(" ");
  const result = await addCategory(name, slug);

  if (result.success) {
    await ctx.reply(`✅ Added category <b>${escapeHtml(name)}</b> (<code>${escapeHtml(slug)}</code>)`, {
      parse_mode: "HTML",
    });
    return;
  }

  await ctx.reply(`❌ ${String(result.error ?? "Error")}`);
}

async function cmdStart(ctx: Context): Promise<void> {
  await ctx.reply(
    [
      "📚 <b>Curator Board bot</b>",
      "",
      "Send any URL to save it.",
      "",
      "/list [category] — recent links",
      "/search <query> — search",
      "/delete <id> — remove",
      "/categories — list categories",
      "/addcategory <slug> <Name> — add category",
    ].join("\n"),
    { parse_mode: "HTML" },
  );
}

export function buildBot(token: string): Bot {
  const bot = new Bot(token);
  bot.command("start", cmdStart);
  bot.command("list", cmdList);
  bot.command("delete", cmdDelete);
  bot.command("search", cmdSearch);
  bot.command("categories", cmdCategories);
  bot.command("addcategory", cmdAddCategory);
  bot.on("message:text", async (ctx) => {
    if (ctx.message.text.startsWith("/")) {
      return;
    }
    await handleUrl(ctx);
  });
  return bot;
}
